File: CSharpCodeBase/components/services/daytime.lua
local Class = require 'utils.hump.class'

local DayTime = Class({init=function(self)
      
    self.PHASE_MORNING = 1
    self.PHASE_DAY_START = 2
    self.PHASE_DAY_PROGRESS = 3
    self.PHASE_DAY_END = 4
    self.PHASE_EVENING = 5
    self.PHASE_NIGHT = 6
    local configPhases = {'morning', 'daystart', 'day', 'dayend', 'evening', 'night'}
    
    self.currentDay = 0
    self.phaseTimes = {}
    self.phaseColors = {}
    self.phaseAmbientColors = {}
    
    for k, v in pairs(configPhases) do
      self.phaseTimes[k] = GameController.database:Get('daytime', v .. '/time')
      self.phaseColors[k] = GameController.database:Get('daytime', v .. '/color')
      self.phaseAmbientColors[k] = GameController.database:Get('daytime', v .. '/acolor')
    end
    self.currentPhase = self.PHASE_MORNING
    self.currentPhaseTime = self.phaseTimes[self.currentPhase]
    self.currentPhaseTimeRunning = 0
    self:OnLoadPhase(self.currentPhase)
end})

function DayTime:Update()
  self.currentPhaseTimeRunning = self.currentPhaseTimeRunning + GameController.deltaTime
  self.currentPhaseTime = self.currentPhaseTime - GameController.deltaTime
  if self.currentPhaseTime <= 0 then
    self:SwitchPhase()
  end
end

function DayTime:SwitchPhase()
  local previousPhase = self.currentPhase
  self.currentPhase = self.currentPhase + 1
  if self.currentPhase > self.PHASE_NIGHT then
    self.currentPhase = self.PHASE_MORNING
    self.currentDay = self.currentDay + 1
  end
  self.currentPhaseTime = self.phaseTimes[self.currentPhase]
  self.currentPhaseTimeRunning = 0
  self:OnUnloadPhase(previousPhase)
  self:OnLoadPhase(self.currentPhase)
end

function DayTime:OnUnloadPhase(phase)
  
end

function DayTime:OnLoadPhase(phase)
  print("Current phase " .. phase)
  GameController.eventSystem:Event('DAY_PHASE_CHANGED',
    {
      current = phase,
      color=self.phaseColors[phase],
      ambientColor=self.phaseAmbientColors[phase],
    })
end

return DayTime

File: CSharpCodeBase/components/services/daytime.cs
using System.Collections.Generic;
using System.Runtime.InteropServices;
using UnityEngine;

namespace MainGame
{
    public class daytime
    {
        public enum ePhase
        {
            PHASE_MORNING = 0,
            PHASE_DAY_START = 1,
            PHASE_DAY_PROGRESS = 2,
            PHASE_DAY_END = 3,
            PHASE_EVENING = 5,
            PHASE_NIGHT = 6,
        };

        public ePhase currentPhase;

        public int currentDay = 0;
        private Dictionary<ePhase, float> phaseTimes;
        private Dictionary<ePhase, Color> phaseColors;
        private Dictionary<ePhase, Color> phaseAmbientColors;

        private float currentPhaseTime;
        private float currentPhaseTimeRunning;

        public daytime()
        {
            string[] configPhases = new string[] {"morning", "daystart", "day", "dayend", "evening", "night"};
            currentDay = 0;

            phaseTimes = new Dictionary<ePhase, float>();
            phaseColors = new Dictionary<ePhase, Color>();
            phaseAmbientColors = new Dictionary<ePhase, Color>();

            for (int i = 0; i < configPhases.Length; i++)
            {
                phaseTimes[(ePhase) i] = GameController.database.Get("daytime", configPhases[i] + "/time");
                phaseColors[(ePhase) i] = GameController.database.Get("daytime", configPhases[i] + "/color");
                phaseAmbientColors[(ePhase) i] = GameController.database.Get("daytime", configPhases[i] + "/acolor");
            }

            currentPhase = ePhase.PHASE_MORNING;
            currentPhaseTime = phaseTimes[currentPhase];
            currentPhaseTimeRunning = 0;
            OnLoadPhase(currentPhase);
        }

        public void Update()
        {
            currentPhaseTimeRunning += GameController.deltaTime;
            currentPhaseTime = currentPhaseTime - GameController.deltaTime;
            if (currentPhaseTime <= 0)
            {
                SwitchPhase();
            }
        }

        public void SwitchPhase()
        {
            int currentPhaseInt = (int) currentPhase;
            ePhase previousPhase = currentPhase;
            currentPhaseInt++;
            if (currentPhaseInt > (int) ePhase.PHASE_NIGHT)
            {
                currentPhaseInt = (int)ePhase.PHASE_MORNING;
                currentDay++;
            }
            currentPhase = (ePhase) currentPhaseInt;
            currentPhaseTime = phaseTimes[currentPhase];
            currentPhaseTimeRunning = 0;

            OnUnloadPhase(previousPhase);
            OnLoadPhase(currentPhase);
        }

        public void OnUnloadPhase(ePhase phase)
        {

        }

        public void OnLoadPhase(ePhase phase)
        {
            // FIXME
            // DAY PHASE CHANGED EVENT
            /*
                public void OnLoadPhase(phase){
                    print("Current phase " .. phase);
                    GameController.eventSystem:Event("DAY_PHASE_CHANGED",;
                    {
                    current = phase,;
                    color=this.phaseColors[phase],;
                    ambientColor=this.phaseAmbientColors[phase],;
                    })
             */
        }
    }
}